desktop/ui/src/main/java/org/datacleaner/user/MutableServerInformationCatalog.java

```diff
@@ -0,0 +1,135 @@
+/**
+ * DataCleaner (community edition)
+ * Copyright (C) 2014 Neopost - Customer Information Management
+ *
+ * This copyrighted material is made available to anyone wishing to use, modify,
+ * copy, or redistribute it subject to the terms and conditions of the GNU
+ * Lesser General Public License, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this distribution; if not, write to:
+ * Free Software Foundation, Inc.
+ * 51 Franklin Street, Fifth Floor
+ * Boston, MA  02110-1301  USA
+ */
+package org.datacleaner.user;
+
+import org.datacleaner.configuration.DomConfigurationWriter;
+import org.datacleaner.configuration.ServerInformation;
+import org.datacleaner.configuration.ServerInformationCatalog;
+import org.datacleaner.util.StringUtils;
+
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Mutable/modifyable implementation of the {@link ServerInformationCatalog} interface. Used to
+ * allow the user to change the catalog of ServerInformation at runtime. This server information
+ * catalog wraps an immutable instance, which typically represents what is
+ * configured in datacleaner's xml file.
+ */
+
+public class MutableServerInformationCatalog implements ServerInformationCatalog {
+    private final DomConfigurationWriter _configurationWriter;
+    private final List<ServerInformationChangeListener> _listeners = new LinkedList<>();
+
+    private final UserPreferences _userPreferences;
+
+    public MutableServerInformationCatalog(ServerInformationCatalog immutableDelegate, DomConfigurationWriter configurationWriter, UserPreferences userPreferences) {
+        _configurationWriter = configurationWriter;
+        _userPreferences = userPreferences;
+        String[] serverNames = immutableDelegate.getServerNames();
+
+        for (String name : serverNames) {
+            if (containsServer(name)) {
+                // remove any copies of the server - the immutable (XML)
+                // version should always win
+                removeServer(getServer(name), false);
+            }
+            addServerInformation(immutableDelegate.getServer(name), false);
+        }
+
+    }
+
+    public void removeServer(ServerInformation ds) {
+        removeServer(ds, true);
+    }
+
+    private synchronized void removeServer(ServerInformation serverInformation, boolean externalize) {
+        final List<ServerInformation> userServers = _userPreferences.getUserServers();
```

Comment:
I see where this is coming from ... But unfortunately that origin isn't a good source ;-) I've been wanting to get rid of "UserDatastores" etc. since they circumvent conf.xml. Rather I think we want to _always_ be able to offload this to conf.xml. It's a reminiscence of the days where "xml is difficult, serialization is easier" ruled.